### ISSUE TITLE
Risk reward management

### ISSUE THREAD

**Issue by sylend07** [freqtrade/freqtrade#8781]
## Describe your environment

Operating system: MACOS Ventura 13.4
Python Version: 3.8.9
CCXT version: _____ (pip freeze | grep ccxt)
Freqtrade Version: 2023.4
Running on Docker

  
## Your question
I have a strategy. In my strategy, the stop price is drawn from the ATR value. Take profit will be realized as risk/reward 2. However, I want the amount to be lost when a stop is made in the opened transaction, to always be a certain amount. (for example 50 USDT) So position size needs to be adjusted accordingly in every transaction. How can I do this? I would be glad if you could guide me.

**Comment by dobremha** [comment]
I hope you can understand the idea, self.custom_sl_value_every_trade is your 50 USDT stop.

    def custom_stake_amount(self, pair: str, current_time: datetime, current_rate: float,
                            proposed_stake: float, min_stake: float, max_stake: float,
                            leverage: float, entry_tag: str, side: str,
                            **kwargs) -> float:
        temp_df, _ = self.dp.get_analyzed_dataframe(pair=pair, timeframe=self.timeframe)
        last_candle = temp_df.iloc[-1].squeeze()
        if last_candle['enter_long'] ==1 and last_candle['close'] < last_candle['open']:
            proposed_stake = round(self.custom_sl_value_every_trade / min(1-last_candle['low'] / (last_candle['close'] * self.custom_pair_values[pair]['price_reduction_procent']), 0.005)) / leverage
            logger.info(f">> Proposed Trade Stake: {proposed_stake}")
            return proposed_stake
        elif last_candle['enter_long'] ==1 and last_candle['close'] >= last_candle['open']:
            proposed_stake = round(self.custom_sl_value_every_trade / min(1-last_candle['low'] / (last_candle['open'] * self.custom_pair_values[pair]['price_reduction_procent']), 0.005)) / leverage
            logger.info(f">> Proposed Trade Stake: {proposed_stake}")
            return proposed_stake
        else:
            logger.info(f">> Proposed Stake: {proposed_stake}")
            return proposed_stake

**Comment by dobremha** [comment]
and this, not to get some huge stake that you don't want :

    def confirm_trade_entry(self, pair: str, order_type: str, amount: float, rate: float,
                            time_in_force: str, current_time: datetime, entry_tag: str,
                            side: str, **kwargs) -> bool:
        logger.info(f">> CONF.TRD.ENTRY - Proposed Stake: {amount * rate / self.custom_levier}")
        if amount * rate / self.custom_levier <= self.custom_max_USD_entry:
            return True
        else:
            return False

**Comment by sylend07** [comment]
Thanks for your reply. 
If I fail, can I ask for help again?

**Comment by xmatthias** [comment]
Closing this as i think your question has been answered.

Feel free to comment below / reopen if it's still unclear.

**Comment by sylend07** [comment]
> I hope you can understand the idea, self.custom_sl_value_every_trade is your 50 USDT stop.
> 
> ```
> def custom_stake_amount(self, pair: str, current_time: datetime, current_rate: float,
>                         proposed_stake: float, min_stake: float, max_stake: float,
>                         leverage: float, entry_tag: str, side: str,
>                         **kwargs) -> float:
>     temp_df, _ = self.dp.get_analyzed_dataframe(pair=pair, timeframe=self.timeframe)
>     last_candle = temp_df.iloc[-1].squeeze()
>     if last_candle['enter_long'] ==1 and last_candle['close'] < last_candle['open']:
>         proposed_stake = round(self.custom_sl_value_every_trade / min(1-last_candle['low'] / (last_candle['close'] * self.custom_pair_values[pair]['price_reduction_procent']), 0.005)) / leverage
>         logger.info(f">> Proposed Trade Stake: {proposed_stake}")
>         return proposed_stake
>     elif last_candle['enter_long'] ==1 and last_candle['close'] >= last_candle['open']:
>         proposed_stake = round(self.custom_sl_value_every_trade / min(1-last_candle['low'] / (last_candle['open'] * self.custom_pair_values[pair]['price_reduction_procent']), 0.005)) / leverage
>         logger.info(f">> Proposed Trade Stake: {proposed_stake}")
>         return proposed_stake
>     else:
>         logger.info(f">> Proposed Stake: {proposed_stake}")
>         return proposed_stake
> ```



 "last_candle['close']< last_candle['open']"

What is your logic for using the above code? Is it related to your strategy or is it related to the question I asked?

      proposed_stake = round(self.custom_sl_value_every_trade / min(1-last_candle['low'] / (last_candle['close'] * self.custom_pair_values[pair]['price_reduction_procent']), 0.005)) / leverage

Can you explain the calculation logic here?

What did we use your codes for?  (last_candle['low'] / (last_candle['close'])

**Comment by dobremha** [comment]
Is it related to my strategy, not to your question.

**Comment by sandervdhave** [comment]
Hi,
Thanks for these blocks of code. I am implementing this in one of my strategies (seen below). Can i ask two questions?

1. Why do you calculate: min(1-last_candle['low'] / (last_candle['close']. Can't we just use the 'close'?
2. I don't really understand the self.custom_levier. Can someone elaborate on this? Is there a leverage callback where the leverage is calculated and used in this confirm_trade_entry?

``` python
    def custom_stake_amount(self, pair: str, current_time: datetime, current_rate: float,
                        proposed_stake: float, min_stake: float, max_stake: float,
                        leverage: float, entry_tag: str, side: str,
                        **kwargs) -> float:
        
        # Setting the custom SL value
        self.custom_sl_value_every_trade = 10 # Dollar per trade


        temp_df, _ = self.dp.get_analyzed_dataframe(pair=pair, timeframe=self.timeframe)
        last_candle = temp_df.iloc[-1].squeeze()
        if last_candle['enter_long'] ==1:
            proposed_stake = round(self.custom_sl_value_every_trade / min(1-last_candle['low'] / (last_candle['close'] * self.custom_pair_values[pair]['price_reduction_procent']), 0.005)) / leverage
            logger.info(f">> Proposed Trade Stake: {proposed_stake}")
            return proposed_stake
        
        elif last_candle['enter_short'] ==1:
            proposed_stake = round(self.custom_sl_value_every_trade / min(1-last_candle['low'] / (last_candle['close'] * self.custom_pair_values[pair]['price_reduction_procent']), 0.005)) / leverage
            logger.info(f">> Proposed Trade Stake: {proposed_stake}")
            return proposed_stake

        else:
            logger.info(f">> Proposed Stake: {proposed_stake}")

        return proposed_stake
    
    def confirm_trade_entry(self, pair: str, order_type: str, amount: float, rate: float,
                        time_in_force: str, current_time: datetime, entry_tag: str,
                        side: str, **kwargs) -> bool:
        logger.info(f">> CONF.TRD.ENTRY - Proposed Stake: {amount * rate / self.custom_levier}")
        if amount * rate / self.custom_levier <= self.custom_max_USD_entry:
            return True
        else:
            return False
```

**Comment by dobremha** [comment]
Hi, self.custom_levier is the leverage used in the trade, 2x, 5x, 10x and so on.

**Comment by sandervdhave** [comment]
Hi, massive thanks for a fast reply. 
I still keep wondering how the calculation works:

`proposed_stake = round(self.custom_sl_value_every_trade / min(1-last_candle['low'] / (last_candle['close'] * self.custom_pair_values[pair]['price_reduction_procent']), 0.005)) / leverage`

I kindly ask you to elaborate on this, since i have only 6 month experience with coding and still learning a lot. Is it not easier to simplify this calculation with the one in this movieclip: https://youtu.be/-UJHObtnp5A?t=103 

Big tank you in advance!

**Comment by stash86** [comment]
There is no one single formula for RR management. You can use whatever you want (and what you understand better).
Any more question to dobremha can be done privately. Thanks

**Comment by mhgutier** [comment]
have you figured this out? i use combination of custom stake_amount, custom stoploss and custom exit.

1. use long_sl and short_sl depending on your trading style
2. custom_stop should only trigger once,
3. the custom stake amount will be computed based on your risk
4. custom_stop is 2x the stoploss pct.

```
    stoploss = -1.0
    use_custom_stoploss=True
    use_exit_signal = True

    def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:

        # dataframe['atr'] = pta.atr(dataframe['high'], dataframe['low'], dataframe['close'], length=14)
        dataframe[['long', 'short', 'af','reversal']] = pta.psar(dataframe['ha_high'], dataframe['ha_low'], close=None, af0=None, af=None, max_af=None, offset=None)
     

        # dataframe['long_sl']=dataframe['close'] - (2 * dataframe['atr'].rolling(10).mean())
        # dataframe['short_sl']=dataframe['close'] + (2 * dataframe['atr'].rolling(10).mean())
        dataframe['long_sl']=dataframe['close'] - (2 * dataframe['atr'])
        dataframe['short_sl']=dataframe['close'] + (2 * dataframe['atr'])
        # dataframe['long_sl']=dataframe['BB_lower']
        # dataframe['short_sl']=dataframe['BB_upper']
        # dataframe['long_sl']=dataframe['low'].rolling(6).min()
        # dataframe['short_sl']=dataframe['high'].rolling(6).max()  
        # dataframe['long_sl']=dataframe['pmax']
        # dataframe['short_sl']=dataframe['pmax']
        # dataframe['long_sl']=dataframe['pivot_up'] - dataframe['atr']
        # dataframe['short_sl']=dataframe['pivot_dn'] + dataframe['atr']
        # dataframe['long_sl']=dataframe['low'].rolling(3).min()
        # dataframe['short_sl']=dataframe['high'].rolling(3).max()
        # dataframe['long_sl']=dataframe['low'].rolling(6).min()
        # dataframe['short_sl']=dataframe['high'].rolling(6).max()
        # dataframe['long_sl']=dataframe['RESISTANCE2'] - dataframe['atr']
        # dataframe['short_sl']=dataframe['SUPPORT2']  + dataframe['atr']
        # dataframe['long_sl']=dataframe['ha_open'] - (2 * dataframe['atr'])
        # dataframe['short_sl']=dataframe['ha_open'] + (2 * dataframe['atr'])
        
        return dataframe

    def custom_stake_amount(self, pair: str, current_time: datetime, current_rate: float,
                            proposed_stake: float, min_stake: Optional[float], max_stake: float,
                            leverage: float, entry_tag: Optional[str], side: str,
                            **kwargs) -> float:

        dataframe, _ = self.dp.get_analyzed_dataframe(pair=pair, timeframe=self.timeframe)
        last_candle = dataframe.iloc[-1].squeeze()
        risk=50
        # risk= max_stake * 0.02
        last_candle = dataframe.iloc[-1].squeeze()
        if side == 'long':
            cut_rate=(last_candle['long_sl'])
            cut_loss = abs(stoploss_from_absolute(cut_rate,current_rate,False)) * leverage
        elif side == 'short':
            cut_rate=(last_candle['short_sl'])
            cut_loss = abs(stoploss_from_absolute(cut_rate,current_rate,True)) * leverage
        
        new_stake = risk/cut_loss 

        # Use default stake amount.
        return int(new_stake) 
    
    def custom_stoploss(self, pair: str, trade: 'Trade', current_time: datetime,
                        current_rate: float, current_profit: float, **kwargs) -> float:
        dataframe, _ = self.dp.get_analyzed_dataframe(pair, self.timeframe)
        last_candle = dataframe.iloc[-1].squeeze()
        # ##Static ATR STop
        if trade.is_short:
            stop_rate = last_candle['short_sl']
            if (trade.stop_loss_pct == trade.initial_stop_loss_pct):
                sl=stoploss_from_absolute(stop_rate,current_rate,trade.is_short)   
            else:
                sl = -1
                
        else:
            stop_rate = last_candle['long_sl']     
            if (trade.stop_loss_pct == trade.initial_stop_loss_pct) :
                sl=stoploss_from_absolute(stop_rate,current_rate,trade.is_short)  
            else:
                sl = -1
        return sl * trade.leverage
    
    def custom_exit(self, pair: str, trade: 'Trade', current_time: datetime, current_rate: float,
                        current_profit: float, **kwargs):
        dataframe, _ = self.dp.get_analyzed_dataframe(pair, self.timeframe)
        tag = super().custom_exit(pair, trade, current_time, current_rate, current_profit, **kwargs)
        if tag:
            return tag
            
        enter_tag = 'empty'
        if hasattr(trade, 'enter_tag') and trade.enter_tag is not None:
            enter_tag = trade.enter_tag
        enter_tags = enter_tag.split()
        ##1:2
        if trade.is_short:
            if (current_profit > (-2 * trade.stop_loss_pct)) :
                return f'{enter_tag}_TP'
        else:
            if (current_profit > (-2 * trade.stop_loss_pct)) :
                return f'{enter_tag}_TP'  
```

![image](https://github.com/freqtrade/freqtrade/assets/80815775/a43ec602-accf-4e49-973c-0fd757a844a5)
